### DISCUSSION TITLE
Printing qr/barcode

### DISCUSSION THREAD

**Discussion by vonza** [simphonydeveloper/microsblogs#1] [Q&A] [answered]
Hello Friends,
How can we apply PART13 to PART7

**Comment by simphonydeveloper** [comment]
You just need to merge the two codes.

**Reply by vonza** [reply]
Could you please show a sample

**Reply by simphonydeveloper** [reply]
Hey friends, I'm very busy lately, you can see I haven't updated for a long time.

**Reply by vonza** [reply]
Understood sir. If you find could you please look into it

**Reply by simphonydeveloper** [reply]
you can add my tg account.contract me anytime.

**Reply by simphonydeveloper** [reply, marked as answer]
hi friend. you can test the code. i'm ok.
```c#
 string barcode = OpsContext.Check.CheckNumber.ToString("000000");
            byte[] barcodeBytes = Encoding.ASCII.GetBytes(barcode);
            byte[] command = new byte[] { 10, 10, 10, 29, 102, 1, 29, 104, 75, 27, 97, 1, 29, 72, 2, 29, 107, 69, 6 };
            byte[] trails = new byte[] { 27, 97, 48 };
            byte[] data = new byte[command.Length + barcodeBytes.Length + trails.Length];
            command.CopyTo(data, 0);
            barcodeBytes.CopyTo(data, command.Length);
            trails.CopyTo(data, command.Length + barcodeBytes.Length);

            //custom receipt content
            //example：
            //custom header
            OpsCustomReceiptEventArgs args = new OpsCustomReceiptEventArgs()
            {
                TrailerAction = CustomPrintType.Append,
                CustomTrailer = data
            };

            //The items you need to print
            List<int> checkDetailLinks = new List<int>();
            //example:
            //Print the first two items

            for (int i = 0; i < this.OpsContext.CheckDetail.Count; i++)
            {
                if (i == 2)
                    break;
                checkDetailLinks.Add(this.OpsContext.CheckDetail[i].DetailLink);
            }

            var isEnable = this.OpsContext.AskQuestion("enable customerReceipt");
            //Pass 'true' to use the customer receipt printer and format. If set to 'false', the guest check printer is used
            bool customerReceipt = isEnable;

            //your temd
            DbTenderMedia temd = new DbTenderMedia()
            {
                ObjectNumber = 1001
            };

          

            this.OpsContext.PrintReceipt(args, checkDetailLinks, customerReceipt, temd);
            this.OpsContext.ShowMessage("Success");
```

**Reply by vonza** [reply]
Hello my friend. Is it possible to add text value before and after the barcode image on the trailer
Like:
****Start of Loyalty Data****
BarCode Image
****End of Loyalty Data****

**Reply by simphonydeveloper** [reply]
yes

**Reply by vonza** [reply]
I hope you won't mind sharing a sample for the same. I have tried but it did not work for me

**Comment by vonza** [comment]
My dear Friends. This worked for me. I did code for QR Code as shown below
```c#
byte[] qrdatacontent = Encoding.UTF8.GetBytes("This is Qr Data Content from Dll testing App:Futuristic");

byte[] storeQRCode = new byte[] { 0x1D, 0x28, 0x6B, (byte)(qrdatacontent.Length + 3), 0x00, 0x31, 0x50, 0x30 };
byte[] modQRCode = { 0x1D, 0x28, 0x6B, 0x04, 0x00, 0x31, 0x41, 0x32, 0x00 };
byte[] sizeQRCode = new byte[] { 0x1D, 0x28, 0x6B, 0x03, 0x00, 0x31, 0x43, 0x04 };
byte[] printQRCode = new byte[] { 0x1D, 0x28, 0x6B, 0x03, 0x00, 0x31, 0x51, 0x30 };
byte[] centerPrint = new byte[] { 0x1b, 0x61, 0x01 };
byte[] lastbuffer = new byte[] { 27, 97, 48 };
byte[] buffer = new byte[storeQRCode.Length + modQRCode.Length + sizeQRCode.Length + printQRCode.Length + lastbuffer.Length + qrdatacontent.Length + centerPrint.Length];
storeQRCode.CopyTo(buffer, 0);
qrdatacontent.CopyTo(buffer, storeQRCode.Length);
modQRCode.CopyTo(buffer, storeQRCode.Length + qrdatacontent.Length);
sizeQRCode.CopyTo(buffer, storeQRCode.Length + qrdatacontent.Length + modQRCode.Length);
centerPrint.CopyTo(buffer, storeQRCode.Length + qrdatacontent.Length + modQRCode.Length + sizeQRCode.Length);
printQRCode.CopyTo(buffer, storeQRCode.Length + qrdatacontent.Length + modQRCode.Length + sizeQRCode.Length + centerPrint.Length);
lastbuffer.CopyTo(buffer, storeQRCode.Length + qrdatacontent.Length + modQRCode.Length + sizeQRCode.Length + centerPrint.Length + printQRCode.Length);


OpsCustomReceiptEventArgs args = new OpsCustomReceiptEventArgs()
{
    TrailerAction = CustomPrintType.Append,
    CustomTrailer = buffer
};
List<int> checkDetailLinks = new List<int>();

for (int i = 0; i < this.OpsContext.CheckDetail.Count; i++)
{
    checkDetailLinks.Add(this.OpsContext.CheckDetail[i].DetailLink);
}
bool customerReceipt = true;
DbTenderMedia temd = new DbTenderMedia()
{
    ObjectNumber = 902
};
this.OpsContext.PrintReceipt(args, checkDetailLinks, customerReceipt, temd);
```

**Reply by simphonydeveloper** [reply]
That's great.

**Reply by vonza** [reply]
I have one question though. When we use this tender as a SERVICE TOTAL, I want to have the screen exit the open check just like normal SERVICE TOTAL EVENT. Is it possible here
```c#
DbTenderMedia temd = new DbTenderMedia()
{
    ObjectNumber = 902
};
this.OpsContext.PrintReceipt(args, checkDetailLinks, customerReceipt, temd);
```

**Reply by simphonydeveloper** [reply]
I haven't tested it before. If it's not possible, you can use the code to execute ServiceTotal after PrintReceipt is completed.

**Reply by vonza** [reply]
If you have time please check for me, I will also explore on my side how to do it too

**Comment by vonza** [comment]
> hi friend. you can test the code. i'm ok.
> 
> ```cs
>  string barcode = OpsContext.Check.CheckNumber.ToString("000000");
>             byte[] barcodeBytes = Encoding.ASCII.GetBytes(barcode);
>             byte[] command = new byte[] { 10, 10, 10, 29, 102, 1, 29, 104, 75, 27, 97, 1, 29, 72, 2, 29, 107, 69, 6 };
>             byte[] trails = new byte[] { 27, 97, 48 };
>             byte[] data = new byte[command.Length + barcodeBytes.Length + trails.Length];
>             command.CopyTo(data, 0);
>             barcodeBytes.CopyTo(data, command.Length);
>             trails.CopyTo(data, command.Length + barcodeBytes.Length);
> 
>             //custom receipt content
>             //example：
>             //custom header
>             OpsCustomReceiptEventArgs args = new OpsCustomReceiptEventArgs()
>             {
>                 TrailerAction = CustomPrintType.Append,
>                 CustomTrailer = data
>             };
> 
>             //The items you need to print
>             List<int> checkDetailLinks = new List<int>();
>             //example:
>             //Print the first two items
> 
>             for (int i = 0; i < this.OpsContext.CheckDetail.Count; i++)
>             {
>                 if (i == 2)
>                     break;
>                 checkDetailLinks.Add(this.OpsContext.CheckDetail[i].DetailLink);
>             }
> 
>             var isEnable = this.OpsContext.AskQuestion("enable customerReceipt");
>             //Pass 'true' to use the customer receipt printer and format. If set to 'false', the guest check printer is used
>             bool customerReceipt = isEnable;
> 
>             //your temd
>             DbTenderMedia temd = new DbTenderMedia()
>             {
>                 ObjectNumber = 1001
>             };
> 
>           
> 
>             this.OpsContext.PrintReceipt(args, checkDetailLinks, customerReceipt, temd);
>             this.OpsContext.ShowMessage("Success");
> ```
